python/ray/tune/tests/test_sample.py

```diff
@@ -1038,17 +1055,104 @@ def testConvertOptuna(self):
             metric="a",
             mode="max")
 
+        sampler4 = RandomSampler(seed=1234)
+        searcher4 = OptunaSearch(
+            space=optuna_define_by_run,
+            sampler=sampler4,
+            metric="a",
+            mode="max")
+
+        sampler5 = RandomSampler(seed=1234)
+        searcher5 = OptunaSearch(
+            space=optuna_define_by_run_with_constants,
+            sampler=sampler5,
+            metric="a",
+            mode="max")
+
+        config_constant = searcher5.suggest("0")
+        self.assertIn("constant", config_constant)
+        config_constant.pop("constant")
+
+        sampler6 = RandomSampler(seed=1234)
+        searcher6 = OptunaSearch(
+            space=optuna_define_by_run_invalid,
+            sampler=sampler6,
+            metric="a",
+            mode="max")
+
+        with self.assertRaises(TypeError):
+            searcher6.suggest("0")
+
         config1 = searcher1.suggest("0")
         config2 = searcher2.suggest("0")
         config3 = searcher3.suggest("0")
+        config4 = searcher4.suggest("0")
 
         self.assertEqual(config1, config2)
         self.assertEqual(config1, config3)
+        self.assertEqual(config1, config4)
+        self.assertEqual(config1, config_constant)
         self.assertIn(config1["a"], [2, 3, 4])
         self.assertIn(config1["b"]["x"], list(range(5)))
         self.assertLess(1e-4, config1["b"]["z"])
         self.assertLess(config1["b"]["z"], 1e-2)
 
+        def optuna_define_by_run_branching_invalid(ot_trial):
+            # this is invalid because such a dict cannot be
+            # unflattened (will try to assign child dicts to value under "a",
+            # but that will be an int, instead of a dict)
+            a = ot_trial.suggest_categorical("a", [1, 2])
+            if a == 1:
+                ot_trial.suggest_int("a/b", 0, 3)
+                ot_trial.suggest_int("a/first", 2, 8)
+            else:
+                ot_trial.suggest_int("a/b", 4, 10)
+                ot_trial.suggest_uniform("a/second", -0.4, 0.4)
+
+        def optuna_define_by_run_branching(ot_trial):
+            a = ot_trial.suggest_categorical("a", ["1", "2"])
+            if a == "1":
+                ot_trial.suggest_int("a:1/b", 0, 3)
+                ot_trial.suggest_int("a:1/first", 2, 8)
+            else:
+                ot_trial.suggest_int("a:2/b", 4, 10)
+                ot_trial.suggest_uniform("a:2/second", -0.4, 0.4)
```

Comment:
```suggestion
            if a == "1":
                ot_trial.suggest_int("nest/b", 0, 3)
                ot_trial.suggest_int("nest/first", 2, 8)
            else:
                ot_trial.suggest_int("nest/b", 4, 10)
                ot_trial.suggest_uniform("nest/second", -0.4, 0.4)
```
can we use these? So that one variable name is the same in different branches. This works with vanilla optuna.